File: base/src/main/java/me/shetj/base/tools/app/ActivityUtils.java
package me.shetj.base.tools.app;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.Keep;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.widget.Toast;

/**
 * Created by admin on 2017/9/13.
 */
@Keep
public class ActivityUtils {
    public static void openActivity(Activity activity,String scheme){
        Intent in = new Intent(Intent.ACTION_VIEW, Uri.parse(scheme));
        activity.startActivity(in);
    }

    public static void openActivityByPackageName(Activity activity,String ackageName){
        Intent intent = activity.getPackageManager().getLaunchIntentForPackage(ackageName);
        activity.startActivity(intent);
    }

    public static void addFragmentToActivity (@NonNull FragmentManager fragmentManager,
                                              @NonNull Fragment fragment, int frameId) {
        checkNotNull(fragmentManager);
        checkNotNull(fragment);
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        transaction.add(frameId, fragment);
        transaction.commit();
    }

    public static void replaceFragment(@NonNull FragmentManager fragmentManager,
                                       @NonNull Fragment fragment, int frameId){
        FragmentTransaction transaction = fragmentManager.beginTransaction().addToBackStack(null);
        transaction.replace(frameId, fragment);
        transaction.commit();
    }

    //移除fragment
    public void removeFragment(@NonNull FragmentManager fragmentManager) {
        if (fragmentManager.getBackStackEntryCount() > 1) {
            fragmentManager.popBackStack();
        }
    }


    public static <T> T checkNotNull(T reference) {
        if (reference == null) {
            throw new NullPointerException();
        }
        return reference;
    }
    public static <T> T checkNotNull(T reference, @Nullable Object errorMessage) {
        if (reference == null) {
            throw new NullPointerException(String.valueOf(errorMessage));
        }
        return reference;
    }
    public static void showNoImplementText(Context context){
        Toast.makeText(context,"Add more codes to support!", Toast.LENGTH_SHORT).show();
    }
}

File: base/src/main/java/me/shetj/base/tools/app/ActivityUtils.kt
package me.shetj.base.tools.app

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.transition.Explode
import android.transition.Fade
import android.transition.Slide
import android.transition.Transition
import android.view.Gravity
import android.view.View
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityOptionsCompat
import androidx.core.util.Pair
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import java.util.*


fun Context.openActivity(scheme: String){
    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(scheme))
    startActivity(intent)
}

fun Context.openActivityByPackageName(ackageName: String){
    val intent =  packageManager.getLaunchIntentForPackage(ackageName)
    startActivity(intent)
}

fun FragmentManager.addFragmentToActivity(fragment: Fragment, frameId: Int){
    checkNotNull(fragment)
    val transaction = beginTransaction()
    transaction.add(frameId, fragment)
    transaction.commit()
}
fun FragmentManager.removeFragment( ){
    if ( backStackEntryCount > 1) {
        popBackStack()
    }
}

fun FragmentManager.replaceFragment(fragment: Fragment, frameId: Int) {
    val transaction = beginTransaction().addToBackStack(null)
    transaction.replace(frameId, fragment)
    transaction.commit()
}


/**
 * 当你需要当前界面中的某个元素和新界面中的元素有关时，你可以使用这个动画。效果很赞~！
 */
fun Activity.getActivityOptions(sharedCardView: View, TRANSITION_NAME_CARD: String): ActivityOptionsCompat {
    return ActivityOptionsCompat
            .makeSceneTransitionAnimation(this, sharedCardView, TRANSITION_NAME_CARD)
}

/**
 * 让新的Activity从一个小的范围扩大到全屏
 */
fun Activity.getActivityOptions(view: View): ActivityOptionsCompat {
    return ActivityOptionsCompat.makeScaleUpAnimation(view,
            //The View that the new activity is animating from
            view.width / 2, view.height / 2,
            //拉伸开始的坐标
            0, 0)
}

/**
 * 多个元素和新的Activity相关的情况，注意下第二个参数Pair这个键值对后面有...，标明是可以传入多个Pair对象的
 */
fun Activity.getActivityOptions(  vararg arg1: Pair<View, String>): ActivityOptionsCompat {
    return ActivityOptionsCompat.makeSceneTransitionAnimation(this, *arg1)
}

fun Activity.startNewAcitivity( options: ActivityOptionsCompat, activityClass: Class<*>) {
    val intent = Intent(this, activityClass)
    intent.addFlags(Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
    ActivityCompat.startActivity(this, intent, options.toBundle())
}


@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
fun Activity.setEnterTransition(  type: String) {
    when (type) {
        "explode" -> {
            val explode = Explode()
            explode.duration = 500L
            window.enterTransition = explode
        }
        "slide" -> {
            val slide = Slide(Gravity.BOTTOM)
            slide.duration = 500L
            window.enterTransition = slide
        }
        "fade" -> {
            val fade = Fade()
            fade.duration = 500L
            window.enterTransition = fade
        }
        else -> {
        }
    }
}

/**
 * @param slideTransition  = explode(),slide()(),fade
 * @param transition  share view 的transition  一般为changeBound
 */
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
fun Fragment.setEnterTransition(  slideTransition: Objects, transition: Transition) {
    enterTransition = slideTransition
    allowEnterTransitionOverlap = true
    allowReturnTransitionOverlap = true
    sharedElementEnterTransition = transition

}

fun Context.getIdByName (className: String, resName: String): Int {
    val packageName = packageName
    return applicationContext.resources.getIdentifier(resName, className, packageName)
}


